[mlir][Transforms] Support `moveOpBefore`/`After` in dialect conversion

matthias-springer opened the pull request:
Add a new rewrite class for "operation movements". This rewrite class can roll back `moveOpBefore` and `moveOpAfter`.

`RewriterBase::moveOpBefore` and `RewriterBase::moveOpAfter` is no longer virtual. (The dialect conversion can gather all required information for rollbacks from listener notifications.)

joker-eph commented on mlir/lib/Transforms/Utils/DialectConversion.cpp:
```diff
@@ -970,6 +971,54 @@ class BlockTypeConversionAction : public BlockAction {
   void rollback() override;
 };
+
+/// An operation rewrite.
```
Can you expand on the role of the class, the context where it's used?

The "Action" name for this whole section is not great by the way, since the concept of "Actions" is now core to MLIR tracing...

matthias-springer replied:
I renamed the class to `OperationRewrite`. I added more documentation to the superclass `IRRewrite` in the dependent PR.

matthias-springer committed:
[mlir][Transforms] Support `moveOpBefore`/`After` in dialect conversion

Add a new rewrite action for "operation movements". This action can roll back `moveOpBefore` and `moveOpAfter`.

`RewriterBase::moveOpBefore` and `RewriterBase::moveOpAfter` is no longer virtual. (The dialect conversion can gather all required information for rollbacks from listener notifications.)

BEGIN_PUBLIC
No public commit message needed for presubmit.
END_PUBLIC
